ray-operator/controllers/ray/rayjob_controller.go

```diff
@@ -0,0 +1,262 @@
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	"github.com/ray-project/kuberay/ray-operator/controllers/ray/utils"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	rayv1alpha1 "github.com/ray-project/kuberay/ray-operator/apis/ray/v1alpha1"
+)
+
+const (
+	RayJobDefaultRequeueDuration = 3 * time.Second
+)
+
+// RayJobReconciler reconciles a RayJob object
+type RayJobReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	Log    logr.Logger
+}
+
+// NewRayJobReconciler returns a new reconcile.Reconciler
+func NewRayJobReconciler(mgr manager.Manager) *RayJobReconciler {
+	return &RayJobReconciler{
+		Client: mgr.GetClient(),
+		Scheme: mgr.GetScheme(),
+		Log:    ctrl.Log.WithName("controllers").WithName("RayJob"),
+	}
+}
+
+// +kubebuilder:rbac:groups=ray.io,resources=rayjobs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=ray.io,resources=rayjobs/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=ray.io,resources=rayjobs/finalizer,verbs=update
+// +kubebuilder:rbac:groups=core,resources=events,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=pods/status,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=services/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=coordination.k8s.io,resources=leases,verbs=get;list;create;update
+// +kubebuilder:rbac:groups=core,resources=serviceaccounts,verbs=get;list;watch;create;delete
+// +kubebuilder:rbac:groups="rbac.authorization.k8s.io",resources=roles,verbs=get;list;watch;create;delete;update
+// +kubebuilder:rbac:groups="rbac.authorization.k8s.io",resources=rolebindings,verbs=get;list;watch;create;delete
+
+// Reconcile reads that state of a RayJob object and makes changes based on it
+// and what is in the RayJob.Spec
+// Automatically generate RBAC rules to allow the Controller to read and write workloads
+// Reconcile used to bridge the desired state with the current state
+func (r *RayJobReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	_ = r.Log.WithValues("rayjob", request.NamespacedName)
+	log.Info("reconciling RayJob", "NamespacedName", request.NamespacedName)
+
+	// Get RayJob instance
+	var err error
+	var rayJobInstance *rayv1alpha1.RayJob
+
+	if rayJobInstance, err = r.getRayJobInstance(ctx, request); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// Set rayClusterName and rayJobId first, to avoid duplicate submission
+	err = r.setRayJobIdAndRayClusterNameIfNeed(ctx, rayJobInstance)
+	if err != nil {
```

Comment:
@brucez-anyscale 
https://github.com/ray-project/kuberay/pull/303#discussion_r921814738
I now save cluster name and job Id before continue reconciling rayJob. In this case, rayCluster will not be created duplicately because only the first updating of status will succeed

Reply:
If you update status in `setRayJobIdAndRayClusterNameIfNeed`, you should better just return this reconcile loop.
Or you should get `rayJobInstance` again, since the `resource version` of rayJobInstance has changed.